### PR TITLE
Enforce rules for OpTypeRuntimeArray on Vulkan

### LINKED ISSUE #1936: spirv-val doesn't validate Vulkan restrictions for  OpTypeRuntimeArray
Vulkan spec says:

OpTypeRuntimeArray must only be used for:
◦ the last member of an OpTypeStruct that is in the StorageBuffer storage class decorated as
Block, or that is in the Uniform storage class decorated as BufferBlock.

This CTS tests violates it  dEQP-VK.spirv_assembly.instruction.compute.pointer_parameter.workgroup_memory_variable_pointers contains

### PR DESCRIPTION
**Pull request by zoddicus**
Fixes #1936

### CODE REVIEW

File: source/val/validate_memory.cpp [hunk 1]
```diff
@@ -502,6 +502,82 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
     }
   }
 
+  // Vulkan specific validation rules
+  if (spvIsVulkanEnv(_.context()->target_env)) {
+    const auto type_index = 2;
+    const auto value_id = result_type->GetOperandAs<uint32_t>(type_index);
+    auto value_type = _.FindDef(value_id);
+    // OpTypeRuntimeArray should only ever be in an OpTypeStruct, so should
+    // never appear as a bare variable.
```
Comment (jeffbolznv) [inline source/val/validate_memory.cpp]: This looks like it doesn't handle the changes from SPV_EXT_descriptor_indexing, which also allows runtime arrays of descriptor variables. The Vulkan rules currently say:

```
  * code:OpTypeRuntimeArray must: only be used for:
  ** the last member of an code:OpTypeStruct
ifdef::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]
     that is in the code:StorageBuffer storage class decorated as
     code:Block, or
endif::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]
     that is in the code:Uniform storage class decorated as
     code:BufferBlock.
ifdef::VK_EXT_descriptor_indexing[]
  ** If the code:RuntimeDescriptorArrayEXT capability is supported, an array
     of variables with storage class code:Uniform,
ifdef::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]
     code:StorageBuffer,
endif::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]
     or code:UniformConstant, or for the outermost dimension of an array of
     arrays of such variables.
endif::VK_EXT_descriptor_indexing[]
```

Comment (zoddicus) [reply source/val/validate_memory.cpp]: Thanks for pointing this out. I have updated the PR to handle this case.

**Comment by zoddicus** [comment]
I have created tests for the RTA of RTA and Array of RTA cases, and you are correct, they are not being caught correctly. I will add in the other container cases.

**Comment by zoddicus** [comment]
@alan-baker PTAL, I have added in the missing checks and expanded the test coverage.

**Comment by zoddicus** [comment]
Talked with David a bit about this, this afternoon, and I think I have gotten my head around what I need to be checking/what it means. I will revise this PR when I get back from the holiday break.

**Comment by zoddicus** [comment]
@alan-baker @dneto0 PTAL

File: source/val/validate_memory.cpp [hunk 2]
```diff
@@ -197,6 +197,14 @@ bool ContainsInvalidBool(ValidationState_t& _, const Instruction* storage,
   return false;
 }
 
+bool HasDecoration(ValidationState_t& _, uint32_t id, SpvDecoration dec) {
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply source/val/validate_memory.cpp]: Done

File: source/val/validate_memory.cpp [hunk 3]
```diff
@@ -569,6 +569,79 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
...
+          return _.diag(SPV_ERROR_INVALID_ID, inst)
+                 << "For Vulkan with RuntimeDescriptorArrayEXT, an variable "
+                    "containing OpTypeRuntimeArray must have storage class of "
+                    "StorageBuffer, Uniform, or UniformConstant.";
+        }
+      }
+    }
+
+    // If an OpStruct has an OpTypeRuntimeArray somewhere within it, then it
+    // must either have the storage class StorageBuffer and be be decorated
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply source/val/validate_memory.cpp]: Done

File: source/val/validate_memory.cpp [hunk 4]
```diff
@@ -569,6 +569,79 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
...
+                  "appear as the final member of an OpTypeStruct, thus cannot "
+                  "be instantiated via OpVariable";
+      } else {
+        // A bare variable OpTypeRuntimeArray is allowed in this context, but
+        // still need to check the storage class.
+        if (storage_class != SpvStorageClassStorageBuffer &&
+            storage_class != SpvStorageClassUniform &&
+            storage_class != SpvStorageClassUniformConstant) {
+          return _.diag(SPV_ERROR_INVALID_ID, inst)
+                 << "For Vulkan with RuntimeDescriptorArrayEXT, an variable "
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply source/val/validate_memory.cpp]: Done

**Comment by zoddicus** [comment]
PTAL, implemented comments and added tests.

File: test/val/val_memory_test.cpp [hunk 5]
```diff
@@ -1975,6 +1972,35 @@ OpFunctionEnd
...
+  std::string spirv = R"(
+OpCapability RuntimeDescriptorArrayEXT
+OpCapability Shader
+OpExtension "SPV_EXT_descriptor_indexing"
+OpMemoryModel Logical GLSL450
+OpEntryPoint Fragment %func "func"
+OpExecutionMode %func OriginUpperLeft
+OpDecorate %array_t ArrayStride 4
+OpMemberDecorate %struct_t 0 Offset 0
+OpDecorate %struct_t BufferBlock
```
Comment (alan-baker) [inline test/val/val_memory_test.cpp]: This makes the uniform a SSBO. This decoration should be Block for a UBO.

Comment (zoddicus) [reply test/val/val_memory_test.cpp]: Done

**Pull request merged**